### PR TITLE
feature(k8s_local_minikube): create local minikube backend

### PR DESCRIPTION
**Pull request by dkropachev**
https://trello.com/c/8rQRn4F1/3425-make-local-minikube-kind-cluster?menu=filter&filter=member:dmitrykropachev

Depends on the following PRs:
https://github.com/scylladb/scylla-cluster-tests/pull/3610
https://github.com/scylladb/scylla-cluster-tests/pull/3609

## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [ ] ~~New configuration option are added and documented (in `sdcm/sct_config.py`)~~
- [ ] ~~I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)~~
- [x] All new and existing unit tests passed (CI)
- [ ] ~~I have updated the Readme/doc folder accordingly (if needed)~~

### CODE REVIEW

File: defaults/k8s_local_kind_config.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+# NOTE: If 'k8s_scylla_operator_docker_image' not set then the one from helm chart will be used.
+# To test nightly builds define it like this: 'scylladb/scylla-operator:nightly'
+k8s_scylla_operator_docker_image: ''
+k8s_scylla_operator_helm_repo: 'https://storage.googleapis.com/scylla-operator-charts/latest'
+k8s_scylla_operator_chart_version: 'latest'
+k8s_cert_manager_version: '1.2.0'
+
+k8s_scylla_datacenter: 'gce'
+k8s_scylla_rack: 'minikube'
```
Comment (bentsi) [inline defaults/k8s_local_kind_config.yaml:19]: kind

File: defaults/k8s_local_kind_config.yaml [hunk 2]
```diff
@@ -0,0 +1,33 @@
+instance_provision: 'on_demand'
+
+minikube_version: '0.11.0'
```
Comment (bentsi) [inline defaults/k8s_local_kind_config.yaml:3]: Add comment and:
`minikube_version` -> `mini_k8s_version`

Comment (vponomaryov) [reply defaults/k8s_local_kind_config.yaml:3]: Actually "minikube" is the name of project.
https://minikube.sigs.k8s.io/docs/start/

File: defaults/k8s_local_kind_config.yaml (hunk 1 above)
Comment (vponomaryov) [reply defaults/k8s_local_kind_config.yaml:19]: "minikube" and "kind" are names of 2 different projects. So, we must use here the proper project name.

Comment (dkropachev) [reply defaults/k8s_local_kind_config.yaml]: @vponomaryov , yap, that is why we change it to kind in kind's default configuration file

File: defaults/k8s_local_kind_config.yaml (hunk 2 above)
Comment (dkropachev) [reply defaults/k8s_local_kind_config.yaml]: @vponomaryov , you are right, but this is perfect case when we can avoid growing number of sct options by having general option that could be reused by multiple backends

File: defaults/k8s_local_kind_config.yaml (hunk 1 above)
Comment (dkropachev) [reply defaults/k8s_local_kind_config.yaml]: fixed

File: sdcm/cluster_k8s/minikube.py [hunk 3]
```diff
@@ -141,35 +301,159 @@ def local_kubectl_version(self):  # pylint: disable=no-self-use
...
+        return self.host_node.remoter
+
+    def deploy_scylla_manager(self, pool_name: str = None) -> None:
+        self.deploy_minio_s3_backend()
+        super().deploy_scylla_manager(pool_name=pool_name)
+
+    @property
+    @abc.abstractmethod
+    def _create_kubectl_config_cmd(self):
+        pass
```
Comment (vponomaryov) [inline sdcm/cluster_k8s/minikube.py:319]: You may want to use following python3 approach to reduce number of lines:

```
    @property
    @abc.abstractmethod
    def _create_kubectl_config_cmd(self): ...
```

Three dots is syntax sugar which allows us to skip usage of "pass" statement.

File: sdcm/cluster_k8s/minikube.py [hunk 4]
```diff
@@ -141,35 +301,159 @@ def local_kubectl_version(self):  # pylint: disable=no-self-use
...
+
+    def deploy(self):
+        if cluster.Setup.REUSE_CLUSTER:
+            if self.is_docker_installed and \
+                    self.is_kubectl_installed and \
+                    self.is_k8s_software_installed and \
+                    self.is_k8s_software_running:
+                self.create_kubectl_config()
+                return
```
Comment (vponomaryov) [inline sdcm/cluster_k8s/minikube.py:378]: Please, see this: https://www.python.org/dev/peps/pep-0008/

It is better to not use backslashes if possible and use "and/or" operator on new line to show that it is condition continuation:

```
if cluster.Setup.REUSE_CLUSTER:
    if (self.is_docker_installed
            and self.is_kubectl_installed
            and self.is_k8s_software_installed
            and self.is_k8s_software_running):
        self.create_kubectl_config()
        continue
    raise RuntimeError("SCT_REUSE_CLUSTER is set, but target host is not ready")
```

File: sdcm/tester.py [hunk 5]
```diff
@@ -932,11 +985,11 @@ def get_cluster_k8s_gce_minikube(self):
...
-            minikube.MinikubeScyllaPodCluster(k8s_cluster=self.k8s_cluster,
-                                              scylla_cluster_name=self.params.get("k8s_scylla_cluster_name"),
-                                              user_prefix=self.params.get("user_prefix"),
-                                              n_nodes=self.params.get("n_db_nodes"),
-                                              params=self.params)
+            minikube.RemoteMinimalScyllaPodCluster(k8s_cluster=self.k8s_cluster,
+                                                   scylla_cluster_name=self.params.get("k8s_scylla_cluster_name"),
+                                                   user_prefix=self.params.get("user_prefix"),
+                                                   n_nodes=self.params.get("n_db_nodes"),
+                                                   params=self.params)
```
Comment (vponomaryov) [inline sdcm/tester.py:992]: Can you please write it in the following way? - 
```
self.db_cluster = minikube.RemoteMinimalScyllaPodCluster(
    k8s_cluster=self.k8s_cluster,
    scylla_cluster_name=self.params.get("k8s_scylla_cluster_name"),
    user_prefix=self.params.get("user_prefix"),
    n_nodes=self.params.get("n_db_nodes"),
    params=self.params,
)
```

**Review by vponomaryov — CHANGES_REQUESTED** [review]

File: sdcm/cluster_k8s/minikube.py [hunk 6]
```diff
@@ -264,54 +553,86 @@ def hydra_dest_ip(self):
...
+        elif isinstance(self.k8s_cluster, KindK8sMixin):
+            return [
+                'drain_k8s_node',
+                # NOTE: enable below methods when it's support fully implemented
+                # https://trello.com/c/LrAObHPC/3119-fix-gce-node-termination-nemesis-on-k8s
+                # https://github.com/scylladb/scylla-operator/issues/524
+                # https://github.com/scylladb/scylla-operator/issues/507
+                # 'terminate_k8s_host',
+                # 'terminate_k8s_node',
+            ]
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply sdcm/cluster_k8s/minikube.py:577]: I will create task on that

File: defaults/k8s_local_kind_config.yaml [hunk 7]
```diff
@@ -0,0 +1,34 @@
+instance_provision: 'on_demand'
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply defaults/k8s_local_kind_config.yaml:1]: removed

File: defaults/k8s_gce_minikube_config.yaml [hunk 8]
```diff
@@ -7,7 +7,9 @@ gce_image_minikube: 'https://www.googleapis.com/compute/v1/projects/ubuntu-os-cl
 gce_instance_type_minikube: 'n1-highmem-8'
 gce_root_disk_type_minikube: 'pd-ssd'
 gce_root_disk_size_minikube: 120
-minikube_version: '1.11.0'
+
+# Version of the minikube to be used
+mini_k8s_version: '1.11.0'
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply defaults/k8s_gce_minikube_config.yaml:12]: fixed to 1.20.0

File: defaults/k8s_local_kind_config.yaml [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+# Version of the kind to be used
+mini_k8s_version: '0.11.0'
+
+n_db_nodes: 3
+
+scylla_version: '4.2.0'
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply defaults/k8s_local_kind_config.yaml:8]: -> 4.4.1

File: defaults/k8s_local_kind_config.yaml [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+mgmt_docker_image: 'scylladb/scylla-manager:2.3.0'
+
+# NOTE: If 'k8s_scylla_operator_docker_image' not set then the one from helm chart will be used.
+# To test nightly builds define it like this: 'scylladb/scylla-operator:nightly'
+k8s_scylla_operator_docker_image: ''
+k8s_scylla_operator_helm_repo: 'https://storage.googleapis.com/scylla-operator-charts/latest'
+k8s_scylla_operator_chart_version: 'latest'
+k8s_cert_manager_version: '1.2.0'
+
+k8s_scylla_datacenter: 'gce'
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply defaults/k8s_local_kind_config.yaml:19]: fixed

File: sdcm/cluster_k8s/minikube.py [hunk 11]
```diff
@@ -74,29 +77,41 @@ def setup_minikube_ubuntu18(node: cluster.BaseNode, kubectl_version: str, miniku
...
+            """)
+        node.remoter.run(f'sudo bash -cxe "{script}"', change_context=True)
 
+    @classmethod
+    def setup_kubectl(cls, node: cluster.BaseNode, kubectl_version: str) -> None:
+        LOGGER.debug("Prepare %s to run Minikube cluster", node)
+        if node.distro.is_ubuntu:
+            cls.setup_kubectl_ubuntu(node, kubectl_version=kubectl_version)
+        else:
+            raise ValueError(f"{node.distro} is not supported by SCT for running Minikube")
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply sdcm/cluster_k8s/minikube.py:89]: fixed

File: defaults/k8s_local_minikube_config.yaml [hunk 12]
```diff
@@ -0,0 +1,35 @@
+instance_provision: 'on_demand'
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply defaults/k8s_local_minikube_config.yaml:1]: removed

File: defaults/k8s_local_minikube_config.yaml [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+
+mini_k8s_version: '1.11.0'
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply defaults/k8s_local_minikube_config.yaml:3]: fixed

File: defaults/k8s_local_minikube_config.yaml [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+scylla_mgmt_agent_version: '2.3.0'
+mgmt_docker_image: 'scylladb/scylla-manager:2.3.0'
+
+# NOTE: If 'k8s_scylla_operator_docker_image' not set then the one from helm chart will be used.
+# To test nightly builds define it like this: 'scylladb/scylla-operator:nightly'
+k8s_scylla_operator_docker_image: ''
+k8s_scylla_operator_helm_repo: 'https://storage.googleapis.com/scylla-operator-charts/latest'
+k8s_scylla_operator_chart_version: 'latest'
+k8s_cert_manager_version: '1.2.0'
+k8s_deploy_monitoring: false
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply defaults/k8s_local_minikube_config.yaml:17]: For testing purpose

File: docs/configuration_options.md [hunk 15]
```diff
@@ -164,7 +164,7 @@
 | **<a href="#user-content-k8s_scylla_cluster_name" name="k8s_scylla_cluster_name">k8s_scylla_cluster_name</a>**  |  | N/A | SCT_K8S_SCYLLA_CLUSTER_NAME
 | **<a href="#user-content-k8s_scylla_disk_gi" name="k8s_scylla_disk_gi">k8s_scylla_disk_gi</a>**  |  | N/A | SCT_K8S_SCYLLA_DISK_GI
 | **<a href="#user-content-k8s_loader_cluster_name" name="k8s_loader_cluster_name">k8s_loader_cluster_name</a>**  |  | N/A | SCT_K8S_LOADER_CLUSTER_NAME
-| **<a href="#user-content-minikube_version" name="minikube_version">minikube_version</a>**  |  | N/A | SCT_MINIKUBE_VERSION
+| **<a href="#user-content-mini_k8s_version" name="mini_k8s_version">mini_k8s_version</a>**  |  | N/A | SCT_MINI_K8S_VERSION
 | **<a href="#user-content-k8s_cert_manager_version" name="k8s_cert_manager_version">k8s_cert_manager_version</a>**  |  | N/A | SCT_K8S_CERT_MANAGER_VERSION
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply docs/configuration_options.md:168]: I agree, idea is that it is referring to minimal k8s.
managed is also not saying much, really don't have good name for it

File: sdcm/cluster_k8s/minikube.py [hunk 16]
```diff
@@ -74,29 +77,41 @@ def setup_minikube_ubuntu18(node: cluster.BaseNode, kubectl_version: str, miniku
             kernel.core_pattern=/var/lib/scylla/coredumps/%h-%P-%u-%g-%s-%t.core
             EOF
             sysctl --system
+            """)
+        node.remoter.run(f'sudo bash -cxe "{script}"', change_context=True)
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply sdcm/cluster_k8s/minikube.py:81]: fixed

File: sdcm/cluster_k8s/minikube.py [hunk 17]
```diff
@@ -74,29 +77,41 @@ def setup_minikube_ubuntu18(node: cluster.BaseNode, kubectl_version: str, miniku
...
             chmod +x /usr/local/bin/kubectl
-
-            # Download Minikube binary.
-            curl -fsSLo /usr/local/bin/minikube \
-                https://storage.googleapis.com/minikube/releases/v{minikube_version}/minikube-linux-amd64
-            chmod +x /usr/local/bin/minikube
-        """)
-        node.remoter.run(f'sudo bash -cxe "{minikube_setup_script}"', change_context=True)
+            """)
+        node.remoter.run(f'sudo bash -cxe "{script}"', change_context=True)
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply sdcm/cluster_k8s/minikube.py:99]: fixed

File: sdcm/cluster_k8s/minikube.py (hunk 3 above)
Comment (dkropachev) [reply sdcm/cluster_k8s/minikube.py:319]: fixed everywhere in the file

File: sdcm/cluster_k8s/minikube.py (hunk 4 above)
Comment (dkropachev) [reply sdcm/cluster_k8s/minikube.py:378]: fixed

File: sdcm/tester.py (hunk 5 above)
Comment (dkropachev) [reply sdcm/tester.py:992]: fixed

**Comment by bentsi** [comment]
@vponomaryov Can you go over the addressed comment fixes and "Resolve" them? Thanks. When Done approve(or not) the PR.

File: sdcm/cluster_k8s/mini_k8s.py [hunk 18]
```diff
@@ -0,0 +1,624 @@
...
+        def get_proxy_ip_port():
+            return LOCALRUNNER.run("grep -P '^Starting' proxy.log | grep -oP '127.0.0.1:[0-9]+'").stdout
+
+        ip_port = wait_for(get_proxy_ip_port, timeout=15, throw_exc=True)
+        return ip_port.strip().split(':')
+
+    @staticmethod
+    def get_kubectl_proxy(node: cluster.BaseNode) -> [str, int]:
+        LOGGER.debug("Stop any other process listening on kubectl proxy port")
+        node.remoter.run(f"sudo fuser -v4k {KUBECTL_PROXY_PORT}/tcp", ignore_status=True)
```
Comment (vponomaryov) [inline sdcm/cluster_k8s/mini_k8s.py:117]: "sudo" stuff

File: sdcm/cluster_k8s/mini_k8s.py [hunk 19]
```diff
@@ -0,0 +1,624 @@
...
+              - role: worker
+              - role: worker
+              - role: worker
+              - role: worker
+
+            " > /tmp/kind.cluster.yaml
+            kind delete cluster || true
+            kind create cluster --config /tmp/kind.cluster.yaml
+        """)
+        self.host_node.remoter.run(f"sudo -E bash -cxe '{script}'")
```
Comment (vponomaryov) [inline sdcm/cluster_k8s/mini_k8s.py:190]: "sudo" stuff

File: sdcm/cluster_k8s/mini_k8s.py [hunk 20]
```diff
@@ -0,0 +1,624 @@
...
+            minikube start --driver=none --extra-config=apiserver.service-node-port-range=1-65535
+            ip link set docker0 promisc on
+            chmod 777 -R $MINIKUBE_HOME
+            # [ -z "$KUBECONFIG" ] || exit
+            # ABS_KUBECONFIG_DIR=$(dirname $(realpath -m $KUBECONFIG))
+            # ABS_MINIKUBE_DIR=$(realpath -m ~/.minikube)
+            # cp -r $ABS_MINIKUBE_DIR $ABS_KUBECONFIG_DIR
+            # sed -ri "s/${{ABS_MINIKUBE_DIR////\\\\/}}/${{ABS_KUBECONFIG_DIR////\\\\/}}\\/.minikube/g" $ABS_KUBECONFIG_DIR/config
+        """)
+        self.host_node.remoter.run(f"sudo -E bash -cxe '{script}'")
```
Comment (vponomaryov) [inline sdcm/cluster_k8s/mini_k8s.py:238]: "sudo" stuff

File: sdcm/tester.py [hunk 21]
```diff
@@ -931,32 +988,35 @@ def get_cluster_k8s_gce_minikube(self):
...
+            service=services[:1],
+            credentials=self.credentials,
+            gce_instance_type=self.params.get("gce_instance_type_loader"),
+            gce_n_local_ssd=self.params.get("gce_n_local_ssd_disk_loader"),
+            gce_image_username=self.params.get("gce_image_username"),
+            user_prefix=self.params.get("user_prefix"),
+            n_nodes=self.params.get('n_loaders'),
+            params=self.params,
+            gce_datacenter=gce_datacenter,
+        )
```
Comment (vponomaryov) [inline sdcm/tester.py:1017]: In general it looks nicer, yes. But this change is already huge and such changes only complicate review.

**Review by vponomaryov — CHANGES_REQUESTED** [review]

File: sdcm/cluster_k8s/mini_k8s.py (hunk 20 above)
Comment (dkropachev) [reply sdcm/cluster_k8s/mini_k8s.py:238]: sudo("-E ...") does not look pretty

File: sdcm/cluster_k8s/mini_k8s.py (hunk 19 above)
Comment (dkropachev) [reply sdcm/cluster_k8s/mini_k8s.py:190]: sudo("-E ...") does not look pretty

File: sdcm/cluster_k8s/mini_k8s.py (hunk 18 above)
Comment (dkropachev) [reply sdcm/cluster_k8s/mini_k8s.py:117]: fixed

File: sdcm/tester.py (hunk 21 above)
Comment (dkropachev) [reply sdcm/tester.py:1017]: Let's keep it.

Comment (vponomaryov) [reply sdcm/tester.py:1017]: Having "+831" ne lines "−368" removed ones we must reduce it if possible.

Again, it is not ok to bloat already huge change even more. 
Please, move such "nice" enhancements into a separate PR.

**Review by vponomaryov — CHANGES_REQUESTED** [review]
2 main requests:

1) Please move "nice" changes to separate PR so this change could be smaller keeping the only idea - new backend.

2) Add details to commit messages itself about the backends you add. Why there are 2 of them and how it is going to be used and how differs...

File: sdcm/tester.py (hunk 21 above)
Comment (dkropachev) [reply sdcm/tester.py]: done

**Comment by dkropachev** [comment]
> 2 main requests:
> 
> 1. Please move "nice" changes to separate PR so this change could be smaller keeping the only idea - new backend.
> 2. Add details to commit (the one with new backend) message itself about the backends you add. Why there are 2 of them and how it is going to be used and how differs...
> 
> Plus, it requires resolve of merge conflicts. Please, address.

Done.

File: sdcm/tester.py [hunk 22]
```diff
@@ -932,12 +988,12 @@ def get_cluster_k8s_gce_minikube(self):
...
-                                              scylla_cluster_name=self.params.get("k8s_scylla_cluster_name"),
-                                              user_prefix=self.params.get("user_prefix"),
-                                              n_nodes=self.params.get("n_db_nodes"),
-                                              params=self.params)
+        self.db_cluster = minikube.RemoteMinimalScyllaPodCluster(
+            k8s_cluster=self.k8s_cluster,
+            scylla_cluster_name=self.params.get("k8s_scylla_cluster_name"),
+            user_prefix=self.params.get("user_prefix"),
+            n_nodes=self.params.get("n_db_nodes"),
+            params=self.params)
```
Comment (vponomaryov) [inline sdcm/tester.py]: Revert this one too and everything else LGTM.

**Review by vponomaryov — CHANGES_REQUESTED** [review]

File: sdcm/tester.py (hunk 22 above)
Comment (dkropachev) [reply sdcm/tester.py]: Not doable, minikube.MinikubeScyllaPodCluster is much shorter than minikube.RemoteMinimalScyllaPodCluster, as result no matter what you do line indent got changed.

Comment (vponomaryov) [reply sdcm/tester.py]: Ok, haven't noticed it, thanks.

**Review by vponomaryov — DISMISSED** [review]

**Comment by dkropachev** [comment]
Pipeline is failing due to the changed hydra image

**Comment by bentsi** [comment]
Please resolve conflicts

File: defaults/k8s_local_kind_config.yaml (hunk 2 above)
Comment (bentsi) [reply defaults/k8s_local_kind_config.yaml]: @vponomaryov `mini_k8s_version` is a general variable for storing local minikube versions for minikube and kind, we made this decision in order not to have many variables for all of the "mini" K8S backends

**Review by bentsi — APPROVED** [review]

**Pull request merged**